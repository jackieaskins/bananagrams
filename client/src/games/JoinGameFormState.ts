import { useState, MouseEvent } from 'react';
import { useHistory, useParams } from 'react-router-dom';

import { useSocket } from '../socket/SocketContext';
import { SetState } from '../state/types';
import { GameInfo, GameLocationState } from './types';

type JoinGameParams = {
  gameId: string;
};

type JoinGameFormState = {
  error: string;
  isJoiningGame: boolean;
  onSubmit: (event: MouseEvent<HTMLElement>) => void;
  setUsername: SetState<string>;
  username: string;
};

export const useJoinGameForm = (): JoinGameFormState => {
  const { socket } = useSocket();
  const { push } = useHistory();
  const { gameId } = useParams<JoinGameParams>();

  const [username, setUsername] = useState('');
  const [error, setError] = useState('');
  const [isJoiningGame, setIsJoiningGame] = useState(false);

  const onSubmit = (event: MouseEvent<HTMLElement>): void => {
    event.preventDefault();

    setIsJoiningGame(true);

    socket.emit(
      'joinGame',
      { gameId, username },
      (error: Error, gameInfo: GameInfo) => {
        if (error) {
          setError(error.message);
          setIsJoiningGame(false);
        } else {
          const locationState: GameLocationState = {
            isInGame: true,
            gameInfo,
          };

          push(`/game/${gameId}`, locationState);
        }
      }
    );
  };

  return {
    error,
    isJoiningGame,
    onSubmit,
    setUsername,
    username,
  };
};
