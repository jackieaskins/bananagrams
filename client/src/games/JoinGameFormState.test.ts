import { useState } from 'react';

import { useJoinGameForm } from './JoinGameFormState';

const mockSetIsJoiningGame = jest.fn().mockName('setIsJoiningGame');
const mockSetUsername = jest.fn().mockName('setUsername');
const mockSetError = jest.fn().mockName('setError');
const mockJoinGame = jest.fn().mockName('joinGame');

jest.mock('react', () => ({
  useState: jest.fn(),
}));

jest.mock('../SocketContext', () => ({
  useSocket: () => ({
    joinGame: mockJoinGame,
  }),
}));

const gameId = 'gameId';

describe('JoinGameFormState', () => {
  beforeEach(() => {
    useState
      .mockReturnValueOnce([true, mockSetIsJoiningGame])
      .mockReturnValueOnce(['username', mockSetUsername])
      .mockReturnValueOnce(['error', mockSetError]);
  });

  describe('useJoinGameForm', () => {
    test('returns correct values', () => {
      expect(useJoinGameForm({ gameId })).toMatchInlineSnapshot(`
        Object {
          "error": "error",
          "isJoiningGame": true,
          "onSubmit": [Function],
          "setUsername": [MockFunction setUsername],
          "username": "username",
        }
      `);
    });

    describe('onSubmit', () => {
      beforeEach(() => {
        useJoinGameForm({ gameId }).onSubmit();
      });

      test('sets joining game to true', () => {
        expect(mockSetIsJoiningGame).toHaveBeenCalledWith(true);
      });

      test('calls joinGame', () => {
        expect(mockJoinGame).toHaveBeenCalledWith(
          {
            gameId: 'gameId',
            username: 'username',
          },
          expect.any(Function)
        );
      });

      describe('joinGame callback', () => {
        const joinGameCallback = (error) =>
          mockJoinGame.mock.calls[0][1](error);

        test('sets the error message on error', () => {
          const message = 'Error';
          joinGameCallback({ message });

          expect(mockSetError).toHaveBeenCalledWith(message);
        });

        test('sets joining game to false', () => {
          joinGameCallback({ message: 'Error' });

          expect(mockSetIsJoiningGame).toHaveBeenCalledWith(false);
        });
      });
    });
  });
});
