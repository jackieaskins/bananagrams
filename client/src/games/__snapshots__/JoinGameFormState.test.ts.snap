// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useJoinGameForm has correct default state 1`] = `
{
  "error": "",
  "isJoiningGame": false,
  "onSubmit": [Function],
  "setUsername": [MockFunction setUsername],
  "username": "",
}
`;
